### PR TITLE
DROOLS-3829 / DROOLS-3934: [DMN Designer] Included Models - Data Types on read-only mode / Boxed Expressions for new imported DRG Elements

### PR DESCRIPTION
**Pull request by karreiro**
See:
- [DROOLS-3934](https://issues.jboss.org/browse/DROOLS-3934): [DMN Designer] Included Models - Boxed Expressions does not work for imported DRG Elements
- [DROOLS-3829](https://issues.jboss.org/browse/DROOLS-3829): [DMN Designer] Included Models - When users open a DMN file with imported Data Types, they must be able to see those types on read-only mode

---

This PR reverts the [DROOLS-3897](https://issues.jboss.org/browse/DROOLS-3897).

### CODE REVIEW

**Comment by karreiro** [comment]
Jenkins, please retest this.

**Comment by karreiro** [comment]
Jenkins, please retest this.

**Comment by karreiro** [comment]
Thank you for the review @jomarko.

Regarding the issues mentioned above...

- https://issues.jboss.org/browse/DROOLS-3985
I think maybe you had a problem deploying the container, I've tried here, and I could reach the DMN model. So, I would close this one.

- https://issues.jboss.org/browse/DROOLS-3985
This is a **blocker** issue (I'm checking this one right now), but seems not relate to this PR so I think we should not block this PR by this problem.

- https://issues.jboss.org/browse/DROOLS-3984
I see this one as a **major** issue, but we can handle this separately (maybe even on 7.5). Since it does not block the feature like the DROOLS-3985.

**Comment by karreiro** [comment]
@jomarko Actually, about the DROOLS-3985, I don't think we have a big problem.. I tried this body in the request:
```
{
  'model-namespace' => namespace,
  'model-name' => model_name,
  'dmn-context' => {
    'include1' => {
      'Cars Per Hour' => 10
    }
  }
}
```
..and I got the rule successfully executed :-)
(include1 is the name of my imported model).

**Comment by karreiro** [comment]
Thank you @jomarko.

@manstis @danielzhe If you get a chance to look at this before I go away... ;-)

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelper.java [hunk 1]
```diff
@@ -54,6 +55,17 @@
...
     List<ItemDefinition> getImportedItemDefinitions(final Map<Import, Definitions> importDefinitions);
 
+    /**
+     * This method finds the list of {@link ItemDefinition}s for a given <code>namespace</code>.
+     * @param workspaceProject represents the project that will be scanned.
+     * @param modelName is the value used as the prefix for imported {@link ItemDefinition}s.
+     * @param namespace is the namespace of the model that provides the list of {@link ItemDefinition}s.
+     * @return a list of imported {@link ItemDefinition}s.
+     */
+    List<ItemDefinition> getImportedItemDefinitionsByNamespace(final WorkspaceProject workspaceProject,
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/common/DMNMarshallerImportsHelper.java]: The `DMNIncludedModelsServiceImpl` implementation calls an instance of `DMNMarshallerImportsHelper` to get the Item Definitions. I've created the helper as something that be re-used in the backend, while I see the `DMNIncludedModelsServiceImpl` as something that exposes backend features to the client side.

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypesPage.java [hunk 2]
```diff
@@ -118,6 +120,17 @@ public void onLostFocus() {
         flashMessages.hideMessages();
     }
 
+    public void onRefreshDataTypesListWithNewItemDefinitions(final @Observes RefreshDataTypesListEvent refresh) {
```
Comment (manstis) [inline kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypesPage.java]: Is there likely to be an issue when multiple DMN editors are open and the name of an imported model is updated in one editor instance; with the other instance observing the same event? Possibly not, if both editor instances do not have imports with the same name; however I'll give it a test.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java [hunk 3]
```diff
@@ -1128,4 +1150,8 @@ private static void applyBackgroundStyles(final BackgroundSet bgset,
     public DiagramMetadataMarshaller<Metadata> getMetadataMarshaller() {
         return diagramMetadataMarshaller;
     }
-}
\ No newline at end of file
+
+    public org.kie.dmn.api.marshalling.DMNMarshaller getMarshaller() {
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/DMNMarshaller.java]: Fixed! :-)

**Comment by manstis** [comment]
@karreiro Manual test worked well. Good job.

File: kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/DMNIncludedModelsServiceImpl.java [hunk 4]
```diff
@@ -89,4 +118,13 @@ public DMNIncludedModelsServiceImpl(final DMNPathsHelperImpl pathsHelper,
     private List<Path> getPaths(final WorkspaceProject workspaceProject) {
         return pathsHelper.getDiagramsPaths(workspaceProject);
     }
+
+    private org.kie.dmn.api.marshalling.DMNMarshaller getMarshaller() {
+        return dmnMarshaller.getMarshaller();
```
[the remark this answers is not in the document]
Comment (karreiro) [reply kie-wb-common-dmn/kie-wb-common-dmn-backend/src/main/java/org/kie/workbench/common/dmn/backend/editors/included/DMNIncludedModelsServiceImpl.java]: Fixed! :-)

**Comment by karreiro** [comment]
@manstis Thank you! I've applied the refactoring you suggested (`DMNMarshallerProducer`) in a separated commit :-)

File: kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypesPage.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/types/DataTypesPage.java]: Tested and my fear did not transpire. All good from me.

**Comment by manstis** [comment]
Jenkins please test this.

**Comment by manstis** [comment]
@karreiro I think Jenkins is broken atm :-(

**Comment by karreiro** [comment]
Ohh... yeah.. many PRs don't have their build running 😥

**Pull request merged**